### PR TITLE
Fix disembarkation on coast during diagonal move

### LINKED ISSUE #2171: Adventure map, unable to disembark on Coast
In fheroes2, the hero cannot disembark at places considered "Coast".

In fheroes2:

unable to disembark 02

unable to disembark 03

unable to disembark 04

unable to disembark 05

unable to disembark 06

unable to disembark 07

Here is the save file:
[Unable to disembark 01.zip](https://github.com/ihhub/fheroes2/files/5495901/Unable.to.disembark.01.zip)

### PR DESCRIPTION
**Pull request by ihhub**
close #2171

### CODE REVIEW

**Comment by ihhub** [comment]
Hi @idshibanov , I've updated the logic. Could you please take a look at this pull request?

File: src/fheroes2/world/world_pathfinding.cpp [hunk 1]
```diff
@@ -84,26 +85,58 @@ bool World::isValidPath( const int index, const int direction, const int heroCol
...
-        case Direction::TOP_LEFT:
-            if ( !GetTiles( index - mapWidth ).isWater() || !GetTiles( index - 1 ).isWater() )
+        case Direction::TOP_LEFT: {
+            assert( index >= mapWidth + 1 );
+
+            const bool isLeftTileWater = GetTiles( index - 1 ).isWater();
+            const bool isTopTileWater = GetTiles( index - mapWidth ).isWater();
+            const bool isTopLeftTileWater = GetTiles( index - mapWidth - 1 ).isWater();
+
+            if ( isTopLeftTileWater != isLeftTileWater || isTopLeftTileWater != isTopTileWater ) {
```
Comment (idshibanov) [inline src/fheroes2/world/world_pathfinding.cpp:95]: Check if this condition is correct. Function will return false you try to land diagonally and on of the adjacent tiles is water (i.e. not in the corner but on straight piece of land).

Comment (ihhub) [reply src/fheroes2/world/world_pathfinding.cpp:95]: You're right. Let me fix it.

**Comment by ihhub** [comment]
Hi @idshibanov , please take a look again.

**Review by idshibanov — APPROVED** [review]

**Pull request merged**